packages/apps/job-launcher/server/src/modules/job/job.service.ts

```diff
@@ -530,7 +540,140 @@ export class JobService {
+        updatedAt: LessThanOrEqual(fiveMinutesAgo)
+      },
+      {
+        order: {
+            waitUntil: SortDirection.ASC,
+        },
+      }
+    );
+    if (!jobEntity) return;
+
+    const { id, userId, chainId, escrowAddress, fundAmount } = jobEntity
+  
+    const signer = this.web3Service.getSigner(chainId);
+    const escrowClient = await EscrowClient.build(signer);
+    const tokenAddress = await escrowClient.getTokenAddress(escrowAddress);
+    
+    let refundAmount = new Decimal(fundAmount);
+    if (escrowAddress) {
+      refundAmount = await this.calculateRefundAmount(chainId, tokenAddress, escrowAddress);
+    }
+    
+    await this.paymentService.createRefundPayment({ refundAmount: refundAmount.toNumber(), userId: userId, jobId: id });
+  
+    jobEntity.status = JobStatus.CANCELED;
+    await jobEntity.save();
+
+    return true;
+  }
+
   public async cancelCronJob() {
+    const jobEntity = await this.jobRepository.findOne(
+      {
+        status: JobStatus.TO_CANCEL,
+        retriesCount: LessThanOrEqual(JOB_RETRIES_COUNT_THRESHOLD),
+        waitUntil: LessThanOrEqual(new Date()),
+      },
+      {
+        order: {
+            waitUntil: SortDirection.ASC,
+        },
+      }
+    );
+    if (!jobEntity) return;
+
+    if (jobEntity.escrowAddress) {
+      await this.processEscrowCancellation(jobEntity);
+    }
+    await this.notifyWebhook(jobEntity);
+    
+    jobEntity.status = JobStatus.TO_REFUND;
+    await jobEntity.save();
+
+    return true;
+  }
+
+  public async processEscrowCancellation(jobEntity: JobEntity) {
+      const { chainId, escrowAddress } = jobEntity;
+
+      const signer = this.web3Service.getSigner(chainId);
+      const escrowClient = await EscrowClient.build(signer);
+      
+      const escrowStatus = await escrowClient.getStatus(escrowAddress);
+      if (escrowStatus === EscrowStatus.Complete || escrowStatus === EscrowStatus.Paid || escrowStatus === EscrowStatus.Cancelled) {
+          this.logger.log(ErrorEscrow.InvalidStatusCancellation, JobService.name);
+          throw new BadRequestException(ErrorEscrow.InvalidStatusCancellation);
+      }
+
+      const balance = await escrowClient.getBalance(escrowAddress);
+      if (balance.eq(0)) {
+          this.logger.log(ErrorEscrow.InvalidBalanceCancellation, JobService.name);
+          throw new BadRequestException(ErrorEscrow.InvalidBalanceCancellation);
+      }
+
+      await escrowClient.cancel(escrowAddress);
+  }
+
+  public async calculateRefundAmount(chainId: ChainId, tokenAddress: string, escrowAddress: string) {
+      const signer = this.web3Service.getSigner(chainId);
+      const tokenContract: HMToken = HMToken__factory.connect(tokenAddress, signer);
+
+      const currentBlockNumber = await signer.provider.getBlockNumber();
+      const filter = {
+          address: tokenAddress,
+          topics: [ethers.utils.id('Transfer(address,address,uint256)')],
+          fromBlock: currentBlockNumber - 200,
```

Comment:
Better define `fromBlock` threshold of 200 as a constant, and reuse it from `cancelCronJobOld` function. It's now 100 over there, and looks like it's inconsistent.